[RFC] Virtualization support

lorc opened the pull request:
Hello,

This PR includes set of patches needed for experimental virtualization support.

I tested it on qemu_v8 with XEN and two virtual guests and for me it works pretty fine. I able to run xtest from two virtual guests simultaneously without any interference.

I'm aware that no one wants to review 28 commits at once, so I want to discuss high level design and if there will be no objections I'll split this PR into series of smaller PRs with more localized changes.

I already shared my approach to virtualization in OP-TEE in different places, but I'll think it is worth to repeat all of this there.

So, main problem is that we need maintain multiple instances of almost anything - one instance per virtual guest. Obvious approach is to introduce `struct virtual_guest_ctx` and hold there all state: threads,  heap, mutexes, TAs, sessions, states of cryptographic primitives and so on. Basically, every virtually all global variables should be moved to `struct virtual_guest_ctx` and every function should get this struct as a parameter. This is explicit way and it needs huge changes to all parts of OP-TEE.

Another approach  is to switch whole .data and .bss sections for every virtual guest. So, for OP-TEE it will look like it works with only one REE as usual. I choose this approach.

Naturally, not all state of  OP-TEE is virtual guest-bound. Some state  is system-wide. So I chose tern "Kernel" for global part of OP-TEE and  "TEE" for per-vm part. 

I added .kbss and .kdata sections to store kernel state. As kernel part also uses heap, I added second heap with `kmalloc`, `kfree` and friends. To explicitly move global variables to kernel sections I added `__kdata` and `__kbss` macros and tagged all needed variables. 

I added `struct mmu_context` entity. It stores whole MMU context. There is a default context which is always present. This default context is the only one available if `CFG_VIRTUALIZATION` is disabled. When virtualization is enabled you can allocate additional contexts - one for every virtual guest. With enabled virtualization `.data` is mapped as RO in default MMU context. It serves as a template for real per-guest `.data` sections. Also, this ensures that without active virtual guest context no one touches `.data` or `.bss` sections.

I added two fast SMCs, so hypervisor can tell OP-TEE when new virtual guest is created or when it being destroyed. 

Upon entry, `thread_handle_fast_smc` or `thread_handle_std_smc` calls `!virt_set_guest(args->a7)` to map right `.data` and `.bss` section for calling virtual guest. With active guest context it is possible now to call TEE functionality.

With virtualization enabled there are no "boot thread". During boot only kernel runtime is initialized. TEE runtime will be initialized during first standard SMC call from guest.

Currently only LPAE and ARMv8 is supported. Actually, it even will not build for ARMv7. Naturally, I'll  fix.
Open question is access to HW resources (like HW crypto or HW-backed secure storage). Currently this is not supported.

That was the main highlights of my solution.

I'm going to share my test setup later (it consist of QEMU, XEN, Dom0 and DomU). Originally I built it with Yocto, but I want to try buildroot.

I'll be glad for any comments. 


<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

cxdinter commented:
Hi @lorc 
Congratulation! It seems you already made big progress for op-tee virtualization porting work.

Sorry, I cannot give you sufficient suggestion on code level. Because I am not familiar with op-tee core, just a application user by developing CAs and TAs. But, could you please share what's the root cause why      
 cannot run different CA in different guest OS to communicate with TAs in op-tee OS (without virtualization support) ? And which type hypervisor you used?

Thank you in advance.

jenswi-linaro commented:
@cxdinter, please create an issue instead.

lorc commented:
Hi,

I extended build framework, so it can build XEN and two domains for QEMU v8. Thus, anyone can try my virtualization patches. 

Setup instructions is quite similar to standard build:

```
# mkdir optee_virt
# cd optee_virt
# repo init -u https://github.com/lorc/optee-manifest.git -b qemu_v8_xen -m qemu_v8_xen.xml && repo # sync -j8
# cd ./build
# make toolchains
# CFG_VIRTUALIZATION=y make run
```

This will run qemu with XEN inside. XEN will boot dom0. From dom0 you can run `xtest`. Also you can create second domain: `# xl create /domu.cfg`.

To enter its  console: `# xl console DomU` . To switch back to dom0 console - press `Ctrl+]`.

You can run `xtest`  simultaneously in two VMs and see that it works quite fine. Some test will fail due to limited memory, but this is normal, as memory is now divided between two VMs.

jbech-linaro commented:
@lorc , nice! I gave it a try and indeed seems to work (`-jX` when building seems to fail though). Since I have Swedish keyboard sending `ctrl+]` is a hazzle, is it possible to just change it temporarily in my environment by editing some config or source code file?

lorc commented:
@jbech-linaro , this is what documentation says:

> Detach from a console depends on the system / keyboard layout.
> Default: CTRL ']'
> German keyboard : CTRL '5' (Works for Putty and gnome-terminal)
> German keyboard (Mac): CTRL + ALT '6
> Swedish keyboard (Mac with iTerm): CTRL 'å'
> 

I never tied  so I can't tell if it really works.

lorc commented:
I rebased this PR, so there are less patches, because first part was merged.

Also I reworked core_mmu_v7 so it now also supports contexts. I hope, now there will no build issues.

jenswi-linaro commented:
Please pay attention to the checkpatch warnings/errors, most seem to be legit and it's a bit distracting trying to disregard those while reviewing.

jenswi-linaro commented on lib/libutils/isoc/include/malloc.h:
```diff
@@ -81,4 +81,61 @@ void malloc_get_stats(struct malloc_stats *stats);
+
+#ifdef ENABLE_MDBG
+
+void *kmdbg_malloc(const char *fname, int lineno, size_t size);
```
These kmdbg_* functions are not implemented

jenswi-linaro commented on lib/libutils/isoc/bget_malloc.c:
```diff
@@ -214,7 +218,7 @@ void malloc_reset_stats(void)
 				 struct malloc_stats *stats)
 {
-	uint32_t exceptions = malloc_lock(&malloc_ctx);
+	uint32_t exceptions = malloc_lock(ctx);
```
This should be fixed in the previous commit.

jenswi-linaro commented:
Commits "core_self_tests.c: add tests for kmalloc" and "pta_stats: add support for kmalloc stats":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -284,10 +308,12 @@ SECTIONS
 		KEEP(*(.nozi .nozi.*))
 		. = ALIGN(16);
 		__nozi_end = .;
+#ifndef CFG_VIRTUALIZATION
```
What happens with `.nozi_stack.stack_thread`?

jenswi-linaro commented on lib/libutils/ext/include/compiler.h:
```diff
@@ -43,6 +43,13 @@
+#define __kbss		__section(".kbss")
+#define __kdata		__section(".kdata")
+#else  /* CFG_VIRTUALIZATION */
+#define __kbss
```
These two should probably be defined as `__bss` and `__data` respectively.

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -105,9 +105,9 @@ DECLARE_STACK(stack_abt, CFG_TEE_CORE_NB_CORE, STACK_ABT_SIZE, static);
-const void *stack_tmp_export = (uint8_t *)stack_tmp + sizeof(stack_tmp[0]) -
-			       (STACK_TMP_OFFS + STACK_CANARY_SIZE / 2);
-const uint32_t stack_tmp_stride = sizeof(stack_tmp[0]);
+const void *stack_tmp_export = (uint8_t *)stack_tmp +
```
Unrelated change, please remove

jenswi-linaro commented on core/arch/arm/kernel/thread.c:
```diff
@@ -116,31 +116,31 @@ const uint32_t stack_tmp_stride = sizeof(stack_tmp[0]);
+long thread_user_kdata_sp_offset __kbss;
 static uint8_t thread_user_kdata_page[
 	ROUNDUP(sizeof(thread_core_local), SMALL_PAGE_SIZE)]
 	__aligned(SMALL_PAGE_SIZE) __section(".nozi.kdata_page");
```
This should probably be `.knozi.kdata_page` with `CFG_VIRTUALIZATION` = `y`

lorc commented on lib/libutils/isoc/bget_malloc.c:
```diff
@@ -150,71 +132,120 @@ struct malloc_pool {
 	}
 }
 
+static void raw_malloc_reset_stats(struct malloc_ctx *ctx)
```
I think that `gen_` (for "generic") prefix would suit better. I changed this prefix for new function that I implemented, but there are couple of  existing functions with `raw_` prefix, but without parameters like `hdr_size` or `ftr_size`. What to do with them?

lorc commented:
@jenswi-linaro, thank you for review.

I addressed your comments, most of checkpatch warnings. There are left some, but they are caused either by long lines which I can't make shorter or by legacy code (warning about `unsigned` ).

Also I fixed  issues with pager and LPAE/ARM32. 

Sorry for forced push, but this is only way to fix checkpatch warnings.

jforissier commented:
Hi @lorc,

> Sorry for forced push, but this is only way to fix checkpatch warnings.

Not exactly, because in addition to checking individual patches, Travis runs checkpatch on the whole diff too (this line: https://travis-ci.org/OP-TEE/optee_os/builds/411339613#L1383). So if one patch introduces something bad and a followup patch fixes it, the second checkpatch run should not complain anymore.

The warning about krealloc() can be fixed easily it seems, as well as the line too long with current_ctx (the DMSG we can live with it...).

lorc commented:
Hi @jforissier,

Thank you. Didn't know that Travis runs checkpatch in two ways.

I'll fix  `krealloc()` and I'll also add separate patch that fixes the same issue with `realloc` in the same file.

As for `ret = current_ctx[get_core_pos()]->l1_tables[0][get_core_pos()][user_va_idx];`, checkpatch was unhappy for any break in this  line. I'll try to rewrite it in a more short form.

jforissier commented:
Thanks. A helper variable should do the trick: `pos = get_core_pos();`.

lorc commented:
I made even better: added helper function `get_ctx()` with optimisation for non-virtualized case.

Also I opened separate PR (#2479) for `realloc` and fixed `krealloc` in-place.

jenswi-linaro commented on lib/libutils/isoc/bget_malloc.c:
```diff
@@ -150,71 +132,120 @@ struct malloc_pool {
+static void gen_malloc_get_stats(struct malloc_ctx *ctx,
+				 struct malloc_stats *stats)
+{
+	uint32_t exceptions = malloc_lock(&malloc_ctx);
```
`s/malloc_ctx/ctx/

jenswi-linaro commented:
For "bget_malloc: add kmalloc pool":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
For commit "tee_mm: add TEE_MM_POOL_KMALLOC flag":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
Is it possible to avoid having all memory mapped (MEM_AREA_SEC_RAM_OVERALL)?

lorc commented on core/arch/arm/include/mm/core_mmu.h:
```diff
@@ -575,6 +592,14 @@ void asid_free(unsigned int asid);
+size_t core_mmu_get_total_pages_size(void);
+struct mmu_context *core_alloc_mmu_ctx(void *tables);
+void core_free_mmu_ctx(struct mmu_context *ctx);
+void core_mmu_set_ctx(struct mmu_context *ctx);
```
Sure, I can rename it. Which variant is more suitable?

jenswi-linaro replied:
Thanks. Let's chose partition, abbreviated as prtn then. That should be easy enough to use.

lorc commented:
About `MEM_AREA_SEC_RAM_OVERALL`. This is needed because I need to access banked memory (.bss, pagetables) before virtual guest context is created or outside this context. I think, it is possible to dynamically map needed memory regions. I can add `MEM_AREA_VIRT_VASPACE` and map needed memory there. This can be slower, but probably a bit more secure.

jenswi-linaro commented:
Regarding MEM_AREA_SEC_RAM_OVERALL, let's keep it as it is at least until I understand it better.

jenswi-linaro commented:
For commit `"thread: move threads initialization to one place"`:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -191,19 +191,44 @@
-uint64_t l1_xlation_table[NUM_L1_TABLES][CFG_TEE_CORE_NB_CORE][NUM_L1_ENTRIES]
+
+typedef uint64_t l1_xlat_tbl_t[NUM_L1_ENTRIES];
+typedef l1_xlat_tbl_t l1_xlat_tbls_t[CFG_TEE_CORE_NB_CORE];
```
`l1_xlat_tbl_t` doesn't seem to be used other than here. I'd prefer typedefing a two dimensional array directly instead of using a helper type.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -191,19 +191,44 @@
+	l1_xlat_tbls_t *l1_tables;
+	xlat_tbl_t *xlat_tables;
+	xlat_tbl_t *l2_ta_tables;
+	int xlat_tables_used;
```
Why a signed type for `xlat_tables_used` and `total_xlat_tables`?

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
+	ctx = kmalloc(sizeof(*ctx));
+
+	if (!ctx)
+		return NULL;
```
`asid` leakage?

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
+		return NULL;
+
+	ctx = kmalloc(sizeof(*ctx));
+
```
empty line not needed

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
+struct mmu_context *core_alloc_mmu_ctx(void *tables)
+{
+	struct mmu_context *ctx;
+	char *tbl = tables;
```
minor: I'd prefer `uint8_t` over `char` here as it's not a string.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
+	if (!ctx)
+		return NULL;
+
+	ctx->l1_tables = (l1_xlat_tbls_t *)(void *)tbl;
```
Casting via `void *` isn't needed

lorc replied:
gcc complaints about alignment:

> warning: cast increases required alignment of target type [-Wcast-align]
>  prtn->l1_tables = (l1_xlat_tbls_t *)tbl;

I think, it can't inferre real alignment there.

jenswi-linaro replied:
In that case drop the `(l1_xlat_tbls_t *)` cast instead.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
+		return NULL;
+
+	ctx->l1_tables = (l1_xlat_tbls_t *)(void *)tbl;
+	memset(ctx->l1_tables, 0, SMALL_PAGE_SIZE);
```
We should have a compile type assert that the L1 tables fits in `SMALL_PAGE_SIZE`

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
+	tbl += ROUNDUP(sizeof(l1_xlation_table), SMALL_PAGE_SIZE);
+
+	ctx->xlat_tables = (xlat_tbl_t *)(void *)tbl;
+	memset(ctx->xlat_tables, 0, sizeof(xlat_tables));
```
I find it a bit disturbing to use `sizeof(xlat_tables)` even if it gives the correct size here. I'd prefer the correct expression or perhaps even better a define.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
+	ctx->xlat_tables = (xlat_tbl_t *)(void *)tbl;
+	memset(ctx->xlat_tables, 0, sizeof(xlat_tables));
+	tbl += sizeof(xlat_tables);
+	assert(((vaddr_t)tbl) % SMALL_PAGE_SIZE == 0);
```
If we need an assert like this, then it would probably be more useful to test the that input is properly aligned.

lorc replied:
I'm testing that input is properly aligned. There is a 
`	assert(((vaddr_t)tbl) % SMALL_PAGE_SIZE == 0);`

At the beginning of the function. 

This second assert is to ensure that all math in the function was right so far.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -339,41 +364,96 @@ static paddr_t get_nsec_ddr_max_pa(void)
+
+	ttbr = virt_to_phys(ctx->l1_tables[0][get_core_pos()]);
+
+	write_ttbr0_el1(ttbr | ((paddr_t)ctx->asid << TTBR_ASID_SHIFT));
```
Do we need a `dsb()` before this to make sure that the writes to the tables are visible, or is that guaranteed by other means?

lorc replied:
Writes are guaranteed by existing means. This commit does not add new pagetable entry manipulation. It uses existing functions like `core_mmu_map_region`.

jenswi-linaro commented on core/arch/arm/include/mm/core_mmu.h:
```diff
@@ -367,17 +378,20 @@ struct core_mmu_table_info {
 	unsigned level;
 	unsigned shift;
 	unsigned num_entries;
+	struct mmu_context *ctx;
```
This is not needed when virtualization isn't enabled. There's a few places where there's a static allocation of this struct, so it will carry a cost.

jenswi-linaro commented:
This is such a large and complex pull request that we need to do the review a little different compared to easier pull requests. Please make changes in place in the commits. Where there's a tag apply that and I know that I've got that one covered, avoid making changes to tagged commits after that as it's very likely I'll miss it.

We're using this method on https://github.com/OP-TEE/optee_os/pull/2345 and it seems to work quite well there, but it takes time of course.

Right now I'll take a pause until you've done some updated, then I'll continue reviewing again.

lorc commented:
@jenswi-linaro , I'll do as you suggested. Thank you so much for review. I'll leave note there when I'll address all your comments.

lorc commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -211,6 +211,41 @@ SECTIONS
+		. = ALIGN(16 * 1024);
+		__kheap_end = .;
+	}
+	.knozi (NOLOAD) : {
```
I added the following comment:

```	
        /*
	 * We want to keep all kernel memory in one place, because
	 * it should be always mapped and it is easier to map one
	 * memory region than two.
	 * Next section are NOLOAD ones, but they are followed
	 * by sections with data. Thus, this NOLOAD section will
	 * be included in the resulting binary, filled with zeroes
	 */

```

jenswi-linaro commented on core/arch/arm/mm/core_mmu_v7.c:
```diff
@@ -189,60 +190,89 @@ enum desc_type {
+	.tables_used = 0,
+};
+
+static struct mmu_partition *current_prtn[CFG_TEE_CORE_NB_CORE];
```
Not needed with !CFG_VIRTUALIZATION

jenswi-linaro commented:
For "core_mmu: add get_memory_map() function" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on core/include/kernel/virtualization.h:
```diff
@@ -0,0 +1,24 @@
+
+#define HYP_CLNT_ID 0
+
+uint32_t virt_guest_created(uint16_t guest_id);
```
These functions need some description

jenswi-linaro commented:
For the commit "generic_boot: use kernel allocator during memory discovery" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -449,7 +449,6 @@ static void init_runtime(unsigned long pageable_part __unused)
 	thread_init_boot_thread();
 
 	init_asan();
-	malloc_add_pool(__heap1_start, __heap1_end - __heap1_start);
```
The pager version of `init_runtime()` needs to be updated too.

jenswi-linaro commented:
I'm taking a break reviewing now to give you some time to update commits and apply tags etc. Let me know when I can resume reviewing.

jenswi-linaro commented:
For "core_mmu: add MEM_AREA_SEC_RAM_OVERALL memory type":
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -191,19 +191,56 @@
+	xlat_tbl_t *l2_ta_tables;
+	unsigned int xlat_tables_used;
+	unsigned int total_xlat_tables;
+	int user_va_idx;
```
Isn't this the same for all initialized partitions?

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -191,19 +191,56 @@
+	xlat_tbl_t *xlat_tables;
+	xlat_tbl_t *l2_ta_tables;
+	unsigned int xlat_tables_used;
+	unsigned int total_xlat_tables;
```
Isn't this always `MAX_XLAT_TABLES`?

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -690,26 +845,39 @@ void core_mmu_get_entry_primitive(const void *table, size_t level,
 {
-	return user_va_idx != -1;
+	bool ret;
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
```
For this thread the partition is supposed to remain constant so masking of exceptions shouldn't be needed.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -690,26 +845,39 @@ void core_mmu_get_entry_primitive(const void *table, size_t level,
 void core_mmu_get_user_va_range(vaddr_t *base, size_t *size)
 {
-	assert(user_va_idx != -1);
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
```
Masking not needed

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -191,19 +191,56 @@
+static struct mmu_partition *get_prtn(void)
+{
+#ifdef CFG_VIRTUALIZATION
+	return current_prtn[get_core_pos()];
```
It looks like this function only is called with exceptions masked so it will work.
I think code could be simplified if this function did mask exceptions itself since the returned value of this function will be constant for a particular thread (or even the active partition).

jenswi-linaro commented:
For "thread: move threads initialization to one place"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
I'm taking a break again reviewing now to give you some time to update commits and apply tags etc. Let me know when I can resume reviewing.

lorc commented:
Jens, thank you for review. I'm sorry for the long pause, I was busy with other project and then was on vacation. I'll address you comments till the end of the week.

lorc commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -542,7 +671,11 @@ void core_mmu_set_info_table(struct core_mmu_table_info *tbl_info,
 {
 	vaddr_t va_range_base;
-	void *tbl = xlat_tables_ul1[thread_get_id()];
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
```
Yes, because `thread_get_id()` is extremely unhappy if is exceptions are enabled

lorc replied:
Ah, sorry, please disregard previous comment. Actually `get_core_pos()` was unhappy.

lorc commented:
I addressed comments and rebased changes onto current master branch.

jenswi-linaro commented:
For "generic_boot: move all runtime initialization to init_tee_runtime()" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
For "generic_boot: add virtualization support" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
For commit "entry_fast: enable hypervisor calls":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
For "virt: add virtualization subsystem"
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
For "core_mmu: add core_mmu_init_virtualization() function":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on core/arch/arm/kernel/thread_a64.S:
```diff
@@ -229,13 +229,16 @@ FUNC thread_rpc , :
 	str	x2, [x19, #THREAD_CTX_REGS_SP]
 	ldr	x20, [sp]	/* Get pointer to rv[] */
 	mov	sp, x0		/* Switch to tmp stack */
+	load_wregs x20, 0, 21, 23	/* Load rv[] into w20-w22 */
```
Why is this needed (loading the registers before the call to thread_state_suspend())?

lorc replied:
Because `thread_state_suspend()` calls `virt_unset_guest()` which removes all guest mappings, so we can't read from `rv[]` after that.

jenswi-linaro replied:
Aha, thanks. Yes a comment would be good.

I guess you need the same change in `core/arch/arm/kernel/thread_a32.S`.

lorc replied:
Sure, will add comment. 

As for 32-bit version... Virtualization is not supported on 32 bits right now. I can make similar changes to `thread_a32.S`, but this will be blind shot, because I can't test it with virtualization enabled.

jenswi-linaro commented:
For "core_mmu_lpae: increase number of xlat tables for virtualization" and "tee_mmu: read TA memory ranges from virtualization subsystem":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
I'm done for now.

lorc commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -0,0 +1,386 @@
+	map[entries - 1].va +=
+		(ta_ram - map[entries - 1].va) & CORE_MMU_PGDIR_MASK;
+	map[entries - 1].pa = ta_ram;
+	map[entries - 1].size = TA_RAM_SIZE;
```
Good сatch, thank you!

lorc commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -0,0 +1,386 @@
+	DMSG("TEE RAM: %08" PRIxPA, tee_mm_get_smem(prtn->tee_ram));
+
+	prtn->ta_ram = tee_mm_alloc(&virt_mapper_pool,
+				   TA_RAM_SIZE / VIRT_GUEST_COUNT);
```
Yes, you are right. I wasn't able to allocate memory for the third guest in my setup.
So I introduced helper function to calculate really available memory for a TA RAM:

```
static size_t get_ta_ram_size(void)
{
	return TA_RAM_SIZE / CFG_VIRT_GUEST_COUNT -
		VCORE_UNPG_RW_SZ -
		core_mmu_get_total_pages_size();
}
```

Now, there is no overallocation anymore.

michpappas replied:
This fixes allocation but mappings are still incorrect.

Guest 1:
> D/TC:0 0 configure_guest_prtn_mem:183 TEE RAM: Allocated 0x18f000 bytes at 0x3f200000
D/TC:0 0 configure_guest_prtn_mem:191 TA RAM: Allocated 0x560000 bytes 0x3f38f000
D/TC:0 0 prepare_memory_map:127 New map (3f071000):
D/TC:0 0 prepare_memory_map:133 T: NSEC_SHM         rsz: 00200000, pa: 3ee00000, va: 3aa00000, sz: 00200000 attr: 1031
D/TC:0 0 prepare_memory_map:133 T: SHM_VASPACE      rsz: 00200000, pa: 00000000, va: 3ac00000, sz: 02000000 attr: 0
D/TC:0 0 prepare_memory_map:133 T: RES_VASPACE      rsz: 00200000, pa: 00000000, va: 3ce00000, sz: 00a00000 attr: 0
D/TC:0 0 prepare_memory_map:133 T: IO_NSEC          rsz: 00200000, pa: f8000000, va: 3da00000, sz: 00200000 attr: 31
D/TC:0 0 prepare_memory_map:133 T: IO_NSEC          rsz: 00200000, pa: f7000000, va: 3dc00000, sz: 00200000 attr: 31
D/TC:0 0 prepare_memory_map:133 T: SEC_RAM_OVERALL  rsz: 00200000, pa: 3f000000, va: 3de00000, sz: 01000000 attr: 1831
D/TC:0 0 prepare_memory_map:133 T: TEE_RAM_RX       rsz: 00001000, pa: 3f000000, va: 3f000000, sz: 00041000 attr: 1851
D/TC:0 0 prepare_memory_map:133 T: TEE_RAM_RO       rsz: 00001000, pa: 3f041000, va: 3f041000, sz: 0000f000 attr: 1811
D/TC:0 0 prepare_memory_map:133 T: KERN_RAM_RW      rsz: 00001000, pa: 3f050000, va: 3f050000, sz: 00021000 attr: 1831
D/TC:0 0 prepare_memory_map:133 T: TEE_RAM_RW       rsz: 00001000, pa: 3f200000, va: 3f071000, sz: 0018f000 attr: 1831
D/TC:0 0 prepare_memory_map:133 T: TA_RAM           rsz: 00001000, pa: 3f38f000, va: 3f38f000, sz: 00700000 attr: 1831

Guest 2:
>D/TC:0   configure_guest_prtn_mem:183 TEE RAM: Allocated 0x18f000 bytes at 0x3f900000
D/TC:0   configure_guest_prtn_mem:191 TA RAM: Allocated 0x560000 bytes 0x3fa8f000
D/TC:0   prepare_memory_map:127 New map (3f071000):
D/TC:0   prepare_memory_map:133 T: NSEC_SHM         rsz: 00200000, pa: 3ee00000, va: 3aa00000, sz: 00200000 attr: 1031
D/TC:0   prepare_memory_map:133 T: SHM_VASPACE      rsz: 00200000, pa: 00000000, va: 3ac00000, sz: 02000000 attr: 0
D/TC:0   prepare_memory_map:133 T: RES_VASPACE      rsz: 00200000, pa: 00000000, va: 3ce00000, sz: 00a00000 attr: 0
D/TC:0   prepare_memory_map:133 T: IO_NSEC          rsz: 00200000, pa: f8000000, va: 3da00000, sz: 00200000 attr: 31
D/TC:0   prepare_memory_map:133 T: IO_NSEC          rsz: 00200000, pa: f7000000, va: 3dc00000, sz: 00200000 attr: 31
D/TC:0   prepare_memory_map:133 T: SEC_RAM_OVERALL  rsz: 00200000, pa: 3f000000, va: 3de00000, sz: 01000000 attr: 1831
D/TC:0   prepare_memory_map:133 T: TEE_RAM_RX       rsz: 00001000, pa: 3f000000, va: 3f000000, sz: 00041000 attr: 1851
D/TC:0   prepare_memory_map:133 T: TEE_RAM_RO       rsz: 00001000, pa: 3f041000, va: 3f041000, sz: 0000f000 attr: 1811
D/TC:0   prepare_memory_map:133 T: KERN_RAM_RW      rsz: 00001000, pa: 3f050000, va: 3f050000, sz: 00021000 attr: 1831
D/TC:0   prepare_memory_map:133 T: TEE_RAM_RW       rsz: 00001000, pa: 3f900000, va: 3f071000, sz: 0018f000 attr: 1831
D/TC:0   prepare_memory_map:133 T: TA_RAM           rsz: 00001000, pa: 3fa8f000, va: 3f28f000, sz: 00700000 attr: 1831

Notice above that although memory allocation is now correct, the mappings TEE_RAM and TA_RAM for the two guests are overlapping.

As mentioned in the earlier comment, the allocation changes need to be reflected in prepare_memory_map() too, ie

> map[entries - 1].size = get_ta_ram_size();

lorc commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -384,13 +482,47 @@ void core_init_mmu_tables(struct tee_mmap_region *mm)
+		memcpy(prtn->l1_tables[0][n],
+		       prtn->l1_tables[0][get_core_pos()],
 		       XLAT_ENTRY_SIZE * NUM_L1_ENTRIES);
 	}
```
In my understanding `user_va_idx` is the same for all contexts. So, I initialize it in `core_init_mmu()` and never touch after this.

lorc commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -745,20 +902,20 @@ void core_mmu_set_user_map(struct core_mmu_user_map *map)
 
```
tlbi_all is done at the end of the function. Are you sure, we need to do it earlier?

michpappas replied:
Yeah, both that one and the one at the end of the function are needed. I found out first hand on hikey when mappings looked correct and I'd still get an abort :)

lorc replied:
As you can see, I am not changing that portions of code my patches. 

Maybe you can create a separate PR with your fixes?

lorc commented:
@michpappas, thank you for your review. I addressed most of your comments. As for platform files - I know that platform files are needed to be updated, but I want to do this in other PR, because this one is already too big and I don't want to disturb platform maintainers with all those generic changes.

lorc commented:
@jenswi-linaro, I also addressed your comment regarding `thread_a64.S`.

I made some changes to "virt: add virtualization subsystem", as requested by @michpappas. Can I leave your A-B?

jbech-linaro commented:
@lorc , I believe this is more or less ready to be merged, isn't it? @michpappas what do you say, do you have any additional comment (@lorc suggests to create a new PR for platform related changes as you can see in his comment above).

lorc commented:
Thanks. I just added tags and rebased onto master

lorc commented:
Sorry, I was to fast. This PR now fails smoke test. I need to take a look, what is wrong

jenswi-linaro commented on core/kernel/console.c:
```diff
@@ -96,9 +96,12 @@ void configure_console_from_dt(void)
 
-	stdout_data = strdup(prop->data);
+	/* We can't use strdup() because it tries to use malloc() */
+	stdout_data = kmalloc(strlen(prop->data) + 1);
```
I'd prefer a `kstrdup()`.

lorc replied:
Any suggestion, about where it should be implemented?

e.g. is "isoc/strdup.c" the right place?

jenswi-linaro replied:
No, it should be "ext/kstrdup.c", with prototype in `<string_ext.h>`

I'm not a big fan of this split of our "libc", but I'm afraid that's how it still is.

lorc replied:
Added "kstrdup()" function and reworked patch to use it.

jenswi-linaro commented:
For commits "pl011.c: use kmalloc allocator", "￼libutils: add kstrdup() function", "￼console: use kstrdup() instead of strdup()":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

lorc commented:
Updated tags.

lorc commented:
Hi @jforissier,

I moved `kstdrup` patches earlier.

As for checkpatch, there are three issues currently: 

 -  long `DMESG()` lines  - this is okay, AFAIK 

 -  bare `unsigned max_level` in `core_mmu_find_table()` parameters. I added new parameter, but I didn't touched `unsigned max_level`, it already was present. I can fix this, no problem there. But this fix has nothing to do with virtualization itself... Should I include it into this PR?

-  `COMPILE_TIME_ASSERT(XLAT_TABLES_SIZE == sizeof(xlat_tables)); ` - "Comparisons should place the constant on the right side of the test". Again, I'm not sure what to do. There is a similar check nearby:

`COMPILE_TIME_ASSERT(CORE_MMU_L1_TBL_OFFSET ==  sizeof(l1_xlation_table) / 2)`, which is already present in the code. I can fix this checkpatch warning, but then resulting code will look like this:

```
	COMPILE_TIME_ASSERT(CORE_MMU_L1_TBL_OFFSET ==  sizeof(l1_xlation_table) / 2);
	COMPILE_TIME_ASSERT(sizeof(xlat_tables) == XLAT_TABLES_SIZE);
```

IMHO, this is looking strange, at least.




As for names... "There are only two hard things in Computer Science: cache invalidation and naming things." -- Phil Karlton

I described this in depth in one of the commits. Basically, you can divide OP-TEE code into two parts. I named them "kernel" and "TEE". "Kernel" part is responsible for most basic things: MMU, exception handling, threads, device drivers and so on. "TEE" part is responsible for the TEE logic: TAs, GlobalPlatform APIs, all that stuff.

My approach to virtualization is to have one instance of "kernel" and multiple instances of "TEE". So, you need separate sections for "kernel" and "TEE" memory and separate memory allocators (`kmalloc`/`malloc`). I tried to make virtualization zero-cost, when it is disabled. So, for example, when `CFG_VIRTUALIZATION=n` then `kmalloc` is just an alias for `malloc` and there is no `__kheap_start` because it is not needed. All is working as it worked earlier.

So, returning back to names... Maybe "kernel" is not the right word. But I can't invent better :) But I'm open to suggestions.

jforissier commented:
> I moved `kstdrup` patches earlier.

Great, thanks.

> [about existing code not compliant with checkpatch]

Ah, I had not noticed existing code was already "wrong". Ideally we should have a separate cleanup patch, then you add your stuff afterward. But TBH this PR is big enough and you have spent a lot of time on it already, so I certainly do not feel like bothering you with such trivial things ;-) I'm fine with the current situation.

> As for names... "There are only two hard things in Computer Science: cache invalidation and naming things." -- Phil Karlton

Haha true ;-) Thanks for explaining again, your approach is sane, I am just questioning the name indeed... I don't have a better suggestion at the moment :-/

jforissier commented:
> So, returning back to names... Maybe "kernel" is not the right word. But I can't invent better :) But I'm open to suggestions.

How about "nexus"?

jforissier commented:
@lorc yes for `nmalloc()`.

But let's hear what others think -- @jenswi-linaro @jbech-linaro @etienne-lms (and others!), what do you think?

jbech-linaro commented:
I have no strong opinion, but I'd like to see whatever we end up with being well documented, i.e., say what it is, why it is and where it is supposed to be used.

lorc commented:
Okay, so I replaced "k" prefix with "nex_" prefix everivere. `sed` helped me a lot :)
There are changes to virtually all commits. Should I strip `r-b` tags or can I leave them?

Also, I added patch with documentation. It currently on top of series, for easier review. I'll move it down when I will be reviewed.

jenswi-linaro commented:
Please leave the R-B tags. :-)
Documentation looks good to me:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

jbech-linaro commented on documentation/virtualization.md:
```diff
@@ -0,0 +1,188 @@
+### No compatibility with "normal" mode ###
+
+OP-TEE built with `CFG_VIRTUALIZATION=y` will not work without
+hypervisor, because before executing any standard SMC,
```
s/hypervisor/a hypervisor/.

lorc commented on documentation/virtualization.md:
```diff
@@ -0,0 +1,188 @@
+
+Right only HW that can be used by multiple VMs simultaneously is
+serial console, used for logging. Devices like HW crypto accelerators,
+secure storage devices and others are not supported right now. Drivers
```
Default storage is working just fine. All RPCs are routed to right VMs, so storage  backed by VM will work fine. 

But external devices cannot be used. I'll try to write this more clearly.

lorc commented:
Thank you for review, @jbech-linaro. I addressed all your comments and pushed additional fixup commit.

jbech-linaro commented:
> Thank you for review, @jbech-linaro. I addressed all your comments and pushed additional fixup commit.

Thanks @lorc , feel free to add my R-B to the documentation commits.
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

lorc commented:
@jbech-linaro , @jenswi-linaro  thank for your review. I added tags, moved commit with documentation, fixed checkpatch warnings and rebased onto current master.

jbech-linaro commented:
@lorc , there are actually quite a few comments from me left (unresolved) in the markdown file. Would you mind to check those to see whether you agree with them or not?

Also, we've finally been able to get rid of all temporary hacks/tweaks on IBART this morning. I've restarted all jobs that failed yesterday and all of them succeeded except your PR here. I wonder if it could be that this PR is incompatible with the current IBART build flags (see [here](https://github.com/jbech-linaro/ibart/blob/master/jobdefs/examples/hikey6220.yaml#L37)). I.e., there is some bug showing up when building with pager, rpmb, sdp and so on.

lorc commented:
Oops, sorry :\ Looks like GitHub hidden most of the comments under spoiler (e.g. "show 20 more comments")  and I missed this spoiler in the midst of other comments. I'll address all of them.

jbech-linaro commented:
Don't worry, I've been tricked by that myself many times :)

lorc commented:
I updated documentation and fixed crash on IBART. Crash was caused by mistake during rebase. I added both fixes as a separate patches, so it would be easier to review.

lorc commented:
Strangelly, IBART still fails at test 1013.

jforissier commented:
@lorc moved from #2737 

> However I can see that IBART still fails on xtest 1013 in #2370. But I can't reproduce that with QEMU either with `QEMU_SMP=1` or with `QEMU_SMP=4`. 

I can reproduce it. One CPU gets stuck in an infinite loop in secure world (exception vector probably), I'm investigating with the JTAG debugger. 

And sometimes I get:

```
E/TC:5 1 Core data-abort at address 0x40004000 (translation fault)
E/TC:5 1  esr 0x96000147  ttbr0 0x200003f026140   ttbr1 0x00000000   cidr 0x0
E/TC:5 1  cpu #5          cpsr 0x800003c4
E/TC:5 1  x0  0000000040004000 x1  0000000040005000
E/TC:5 1  x2  0000000000000040 x3  000000000000003f
E/TC:5 1  x4  0000000000000001 x5  0060000000000fc7
E/TC:5 1  x6  0000000000000090 x7  0000000000000000
E/TC:5 1  x8  0000000000000001 x9  000000003f087c90
E/TC:5 1  x10 0000000000000000 x11 0000000000000000
E/TC:5 1  x12 000000003f0389a0 x13 0000000040004f00
E/TC:5 1  x14 00000000400065d5 x15 f0facfcadc045e63
E/TC:5 1  x16 000000003f087cc8 x17 aa12c36100000000
E/TC:5 1  x18 0000000000000000 x19 000000003f017510
E/TC:5 1  x20 000000003f038710 x21 0000000000001a91
E/TC:5 1  x22 0000000040004000 x23 00000000000003b0
E/TC:5 1  x24 000000003f012000 x25 000000003f041000
E/TC:5 1  x26 000000003f012000 x27 0000000000000000
E/TC:5 1  x28 0000000000001801 x29 000000003f02a960
E/TC:5 1  x30 000000003f008964 elr 000000003f001b60
E/TC:5 1  sp_el0 000000003f02a960
E/TC:5 1 Call stack:
E/TC:5 1  0x000000003f001b60 loop_cvac at core/arch/arm/kernel/cache_helpers_a64.S:61
E/TC:5 1 Panic '[abort] abort in abort handler (trap CPU)' at core/arch/arm/kernel/abort.c:559 <get_fault_type>
E/TC:5 1 Call stack:
E/TC:5 1  0x000000003f007ef0 read_pc at core/arch/arm/include/arm64.h:227
```

lorc commented:
@jbech-linaro , sorry, I abused IBART a little to debug that issue with a lock.

I saw suspect log entry:

```
 7510:  D/TC:? 2 __wq_rpc:39 sleep thread 2 0x3f0121a8 -1
  7511:  *D/TC:? 1 __wq_rpc:39 wake  thread 2 0x3f0121a8 -3
  7512:  *D/TC:? 2 thread_rpc_cmd:1581 Thread RPC cmd 4
  7513:  *D/TC:? 1 thread_rpc_cmd:1581 Thread RPC cmd 4
```

Looks like thread 2 tries to suspend self, but thread 1 wakes it up instantly. Is it possible that `OPTEE_RPC_WAIT_QUEUE_WAKEUP` RPC hits NW before `OPTEE_RPC_WAIT_QUEUE_SLEEP` ?

lorc commented:
@jbech-linaro, okay, thanks :)

I added one more patch, that prints timestamp for every trace message. But looks like IBART is offline at the moment. Is there any way, I can get logs with the mentioned patch applied?

jbech-linaro commented:
> But looks like IBART is offline at the moment. Is there any way, I can get logs with the mentioned patch applied?

Sorry, I was about to try enabling https on IBART and then meetings came in-between. But I've now continued that and indeed have enabled https. I'm running your job as a first test using https. So far it it looking good (I should do this kind of maintenance during night time instead ...), so logs should be there in about 5-10 minutes or so. Sorry for the inconvenience.

lorc commented:
> I can reproduce it. One CPU gets stuck in an infinite loop in secure world (exception vector probably), I'm investigating with the JTAG debugger.

@jforissier,  Did you found anything? I can confirm that with this PR, HiKey stucks in secure world forever. But this is not spinlock, apparently. And without HW it is hard to debug it further.

jforissier commented:
>@jforissier, Did you found anything? I can confirm that with this PR, HiKey stucks in secure world forever. But this is not spinlock, apparently. And without HW it is hard to debug it further.

Well, I could not make sense of the PC address when the CPU was stuck. It was outside the address range of the TEE core executable (not looping in a spinlock or exception vector). I will try again.

jforissier commented:
@lorc `git bisect` on your virt_rfc_nex branch points to commit 4fc4681fca13 ("core_mmu: introduce mmu partitions"):

```
$ git bisect log
git bisect start
# bad: [aa6bdeaa346e7784a2d12cbc88722afa1773a011] entry_fast: enable hypervisor calls
git bisect bad aa6bdeaa346e7784a2d12cbc88722afa1773a011
# good: [41b29406e4f54f11f96e3c936329a45d6674ee60] core: syscall_storage_obj_create(): fix a memory leak
git bisect good 41b29406e4f54f11f96e3c936329a45d6674ee60
# bad: [4fc4681fca13e9b251b0feeec218976cfba61348] core_mmu: introduce mmu partitions
git bisect bad 4fc4681fca13e9b251b0feeec218976cfba61348
# good: [9fe5f6d3a9c6482be0251d01f57f4978c23e5aee] virt: tag variables with __nex_data and __nex_bss
git bisect good 9fe5f6d3a9c6482be0251d01f57f4978c23e5aee
# good: [c4faa345b07bfc7c7ad8d6ab049543da76bd58a6] console: use nex_strdup() instead of strdup()
git bisect good c4faa345b07bfc7c7ad8d6ab049543da76bd58a6
# good: [719a71cba9540172b99cda4f9a87c88c1bc7f228] core_mmu: add MEM_AREA_SEC_RAM_OVERALL memory type
git bisect good 719a71cba9540172b99cda4f9a87c88c1bc7f228
# first bad commit: [4fc4681fca13e9b251b0feeec218976cfba61348] core_mmu: introduce mmu partitions
```

Tested with HiKey960, default config (64-bit TEE core, 32-bit TAs) except: `CFG_TEE_CORE_LOG_LEVEL=3 CFG_WITH_PAGER=y CFG_LOCKDEP=y CFG_RPMB_FS=y CFG_RPMB_WRITE_KEY=y` Symptom: `xtest 1013` hangs quite rapidly.

lorc commented:
Thank you @jforissier for your help. I just finished bisecting using IBART. And I got the same culprit. So, I'm glad that our results match. 

It is quite strange, that it jumps outside OP-TEE and hangs there. Anyways, I'll try to figure out what I'm doing wrong in that patch. 

Maybe you have any ideas how HiKey960 differs from other platforms? I tested no QEMU and on Renesas Gen3. It is working quite fine. So, I'm wondering, why this reproduces on HiKey...

lorc commented:
So I managed to run pager on RCAR H3 and managed to reproduce the issues. Actually I can see at least two issues:

1. Data prefetch abort during dcache invalidation.

2. Something changes  value of `VBAR_EL1` to `0x80000800`. In secure mode. I have no idea how this can happen.

I can see that other threads often are blocked at the pager's spinlock. Sometimes I can see that one core is in EL3 mode, and it has absolute garbage in the registers:

![screenshot from 2019-01-29 21-39-56](https://user-images.githubusercontent.com/240174/51935714-2bd50180-240f-11e9-8a0d-448883f3ceb8.png)


My only idea at the moment is that OP-TEE corrupts ARM-TF's memory somehow. So, when ARM-TF resumes core from suspend, it reads garbage from saved context.

But how my patch can cause this?.. Also, it is strange, that we have the very same issue on HiKey, where memory layout is different.

jenswi-linaro commented:
This sounds a bit like `CFG_CORE_UNMAP_CORE_AT_EL0` try building with:
```
 CFG_CORE_UNMAP_CORE_AT_EL0=n
 CFG_CORE_WORKAROUND_SPECTRE_BP=n
 CFG_CORE_WORKAROUND_SPECTRE_BP_SEC=n
```
to make things a bit less complicated.

jenswi-linaro commented:
> Probably, this is the stupidest mistake I ever made.

I know that feeling. :-)

lorc commented:
Squashed and rebased. Travis is not happy about checkpatch, but looks like there is nothing new.

jforissier commented:
@lorc many thanks for the hard work and contributing back to the OSS community. Much appreciated!

jbech-linaro commented:
Congratz @lorc and thanks for having patience with a lengthy review like this!

lorc commented:
Thank you, guys. 

And thank you for your help and support with this patch series.
